shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -77,6 +87,53 @@ typedef NS_ENUM(NSUInteger, FlutterTextAffinity) {
   return flutter::TextRange([base unsignedLongValue], [extent unsignedLongValue]);
 }
 
+// Returns the autofill hint content type, if specified; otherwise nil.
+static NSString* GetAutofillContentType(NSDictionary* configuration) {
+  NSDictionary* autofill = configuration[kAutofillProperties];
+  NSArray<NSString*>* hints = autofill[kAutofillHints];
+  return hints.count > 0 ? hints[0] : nil;
+}
+
+// Returns YES if configuration describes a field for which autocomplete should be enabled for
+// the specified TextInputConfiguration. Autocomplete is enabled by default, but will be disabled
+// if the field is password-related, or if the autofill hint is the empty string.
+static BOOL EnableAutocompleteForTextInputConfiguration(NSDictionary* configuration) {
+  // Disable if obscureText is set.
+  if ([configuration[kSecureTextEntry] boolValue]) {
+    return NO;
+  }
+
+  // Disable if autofill properties indicate a username/password.
+  NSString* contentType = GetAutofillContentType(configuration);
+  if ([contentType isEqualToString:@"password"] || [contentType isEqualToString:@"username"]) {
+    return NO;
+  }
+
+  // Disable if autofill properties specify a blank autofill type.
```

Comment:
When is the autofill type blank? Consider adding a comment, but I don't feel strongly

Reply:
Nice catch. I actually had the same question, spent a bunch of time poking at five layers deep of blame history this morning, and finally got the right answer from @LongCatIsLooong, then promptly forgot to add the doc.

The answer is that while setting obscureText or a content type of password/username disables autofill, sometimes you want to disable it explicitly. This is how we do it.